### PR TITLE
swap parent_root for il_committee_root and introduce timing logic

### PR DESCRIPTION
**Pull request by fradamt**

### CODE REVIEW

File: specs/_features/focil/fork-choice.md [hunk 1]
```diff
@@ -50,44 +56,53 @@ class Store(object):
...
 
 ### New `on_inclusion_list`
 
 `on_inclusion_list` is called to import `signed_inclusion_list` to the fork choice store.
 ```python
 def on_inclusion_list(
-        store: Store, signed_inclusion_list: SignedInclusionList) -> None:
+        store: Store, 
+        signed_inclusion_list: SignedInclusionList, 
+        inclusion_list_committee: Vector[ValidatorIndex, IL_COMMITTEE_SIZE]]) -> None:
```
Comment (jihoonsong) [inline specs/_features/focil/fork-choice.md:69]:
```suggestion
        inclusion_list_committee: List[ValidatorIndex, IL_COMMITTEE_SIZE]]) -> None:
```

File: specs/_features/focil/p2p-interface.md [hunk 2]
```diff
@@ -21,7 +21,7 @@ This document contains the consensus-layer networking specification for FOCIL.
 
 | Name | Value | Unit | Duration |
 | - | - | :-: | :-: |
-| `inclusion_list_CUT_OFF` | `uint64(9)` | seconds | 9 seconds |
+| `attestation_deadline` | `uint64(4)` | seconds | 4 seconds |
```
Comment (jihoonsong) [inline specs/_features/focil/p2p-interface.md:24]: Is this better to be capitalized? Besides, do we want to rename `PROPOSER_INCLUSION_LIST_CUT_OFF` using `DEADLINE` for better consistency?

File: specs/_features/focil/fork-choice.md [hunk 3]
```diff
@@ -50,44 +56,53 @@ class Store(object):
...
 
-    root = hash_tree_root(inclusion_list_committee)
+    is_before_freeze_deadline = get_current_slot(store) == message.slot and time_into_slot < VIEW_FREEZE_DEADLINE
+
+    # Do not process inclusion lists from known equivocators
     if validator_index not in inclusion_list_equivocators[(message.slot, root)]:
         if validator_index in [il.validator_index for il in inclusion_lists[(message.slot, root)]]
             il = [il for il in inclusion_lists[(message.slot, root)] if il.validator_index == validator_index][0]
             if not il == message:
+                # We have equivocation evidence for `validator_index`, record it as equivocator
```
Comment (jihoonsong) [inline specs/_features/focil/fork-choice.md:102]: Do we want to remove `il` from `inclusion_lists[(message.slot, root)]` following the description at L72?

File: specs/_features/focil/fork-choice.md [hunk 4]
```diff
@@ -50,44 +56,53 @@ class Store(object):
...
+        inclusion_list_committee: Vector[ValidatorIndex, IL_COMMITTEE_SIZE]]) -> None:
     """
     ``on_inclusion_list`` verify the inclusion list before importing it to fork choice store.
     If there exists more than 1 inclusion list in store with the same slot and validator index, remove the original one.
     """
     message = signed_inclusion_list.message
-    # Verify inclusion list slot is bounded to the current slot
-    assert get_current_slot(store) == message.slot
+    # Verify inclusion list slot is either from the current or previous slot
+    assert get_current_slot(store) in [message.slot, message.slot + 1]
```
Comment (soispoke) [inline specs/_features/focil/fork-choice.md:76]: wouldn't it be message.slot - 1 if it's from either the current or the previous slot?

File: specs/_features/focil/fork-choice.md [hunk 5]
```diff
@@ -50,44 +56,53 @@ class Store(object):
...
 
-    assert message.beacon_block_root in store.block_states
-    # Get the inclusion list committee for this slot
-    state = copy(store.block_states[message.beacon_block_root])
-    if state.slot < message.slot:
-        process_slots(state, message.slot)
-    inclusion_list_committee = get_inclusion_list_committee(state, message.slot)
+    time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT
+    is_before_attesting_interval = time_into_slot < SECONDS_PER_SLOT // INTERVALS_PER_SLOT
```
Comment (soispoke) [inline specs/_features/focil/fork-choice.md:79]: wouldn't it be easier to define the attestation deadline like this?

    ATTESTATION_DEADLINE = uint64(4)  # seconds
    is_before_attestation_deadline = time_into_slot < ATTESTATION_DEADLINE

    # If the inclusion list is from the previous slot, ignore it if already past the attestation deadline
    if get_current_slot(store) == message.slot + 1:
        assert is_before_attestation_deadline

File: specs/_features/focil/fork-choice.md [hunk 6]
```diff
@@ -50,44 +56,53 @@ class Store(object):
...
+    assert validator_index in inclusion_list_committee
    
     # Verify inclusion list signature
     assert is_valid_inclusion_list_signature(state, signed_inclusion_list)
 
-    root = hash_tree_root(inclusion_list_committee)
+    is_before_freeze_deadline = get_current_slot(store) == message.slot and time_into_slot < VIEW_FREEZE_DEADLINE
+
+    # Do not process inclusion lists from known equivocators
     if validator_index not in inclusion_list_equivocators[(message.slot, root)]:
```
Comment (soispoke) [inline specs/_features/focil/fork-choice.md:98]: I think here you might need to introduce a time window, so you only record equivocation evidence only if before the attestation deadline (t=4s)?

  ```
if validator_index not in inclusion_list_equivocators[(message.slot, root)]:
      if validator_index in [il.validator_index for il in inclusion_lists[(message.slot, root)]]:
          il = next(
              il for il in inclusion_lists[(message.slot, root)] if il.validator_index == validator_index
          )
          if il != message:
              # Record equivocation evidence if within the allowed time window
              if get_current_slot(store) <= message.slot + 1 and (
                  get_current_slot(store) != message.slot + 1 or is_before_attestation_deadline
              ):
                  inclusion_list_equivocators[(message.slot, root)].add(validator_index)
      else:
          # This inclusion list is not an equivocation. Store it if prior to the view freeze deadline
          if is_before_freeze_deadline:
              inclusion_lists[(message.slot, root)].append(message)
```

File: specs/_features/focil/fork-choice.md (hunk 4 above)
Comment (jihoonsong) [reply specs/_features/focil/fork-choice.md:76]: If message is from previous slot, `message.slot + 1 == get_current_slot(store)`.

Comment (soispoke) [reply specs/_features/focil/fork-choice.md:76]: ah right that makes sense :)

File: specs/_features/focil/fork-choice.md (hunk 1 above)
Comment (terencechain) [reply specs/_features/focil/fork-choice.md:69]: I believe Vector is right here

File: specs/_features/focil/fork-choice.md (hunk 3 above)
Comment (terencechain) [reply specs/_features/focil/fork-choice.md:102]: The comment is outdated, as that's the old design. This new design uses `inclusion_list_equivocators` so no longer required to remove `inclusion_lists`

File: specs/_features/focil/fork-choice.md [hunk 7]
```diff
@@ -50,44 +56,53 @@ class Store(object):
...
         if validator_index in [il.validator_index for il in inclusion_lists[(message.slot, root)]]
```
Comment (terencechain) [inline specs/_features/focil/fork-choice.md]:
```suggestion
        if validator_index in [il.validator_index for il in inclusion_lists[(message.slot, root)]]:
```

File: specs/_features/focil/fork-choice.md (hunk 5 above)
Comment (terencechain) [reply specs/_features/focil/fork-choice.md:79]: I dont think we want to hard code `uint64(4)`, we want to use `INTERVALS_PER_SLOT` so it's not a breaking change to the spec if we were to adjust the interval deadlines later

File: specs/_features/focil/fork-choice.md (hunk 1 above)
Comment (jihoonsong) [reply specs/_features/focil/fork-choice.md:69]: What's difference between Vector and List? Is the Vector from Numpy?

Comment (terencechain) [reply specs/_features/focil/fork-choice.md:69]: nope, these are ssz types: https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md
we want a fixed list here

File: specs/_features/focil/fork-choice.md (hunk 6 above)
Comment (fradamt) [reply specs/_features/focil/fork-choice.md:98]: If it's after the attestation deadline, we won't even get here because of line 82

**Pull request merged**